### PR TITLE
drivers: usb: dc: sam: Add atmel USBC controller

### PR DESCRIPTION
**Pull request by nandojve**
Add atmel USBC device controller. ~Fix 'USB High-Speed Interface (USBHS)' driver name by renaming from usb_dc_sam to usb_dc_sam_usbhs. This allows add others SoC drivers like 'USB Device Port (UDP)' that is found at SAM4S/E variations or even 'USBC controller' from SAM4L.~

depends on #36763

Report from Linux testusb:
```
./testusb: /dev/bus/usb/001/094 may see only control tests
/dev/bus/usb/001/094 test 0,    0.000003 secs
/dev/bus/usb/001/094 test 1,    0.943959 secs
/dev/bus/usb/001/094 test 2,    0.891783 secs
/dev/bus/usb/001/094 test 3,    0.943665 secs
/dev/bus/usb/001/094 test 4,    0.892753 secs
/dev/bus/usb/001/094 test 5,   27.003711 secs
/dev/bus/usb/001/094 test 6,   27.029843 secs
/dev/bus/usb/001/094 test 7,   27.069641 secs
/dev/bus/usb/001/094 test 8,   27.075551 secs
/dev/bus/usb/001/094 test 9,    0.756100 secs
/dev/bus/usb/001/094 test 10,    3.057635 secs
/dev/bus/usb/001/094 test 11,   17.292583 secs
/dev/bus/usb/001/094 test 12,   17.599416 secs
/dev/bus/usb/001/094 test 13 --> 22 (error 22)
/dev/bus/usb/001/094 test 14 --> 22 (error 22)
/dev/bus/usb/001/094 test 17,    0.980536 secs
/dev/bus/usb/001/094 test 18,    0.986235 secs
/dev/bus/usb/001/094 test 19,    0.983859 secs
/dev/bus/usb/001/094 test 20,    0.977327 secs
/dev/bus/usb/001/094 test 21 --> 22 (error 22)
/dev/bus/usb/001/094 test 24,   28.330410 secs
/dev/bus/usb/001/094 test 27,   26.947000 secs
/dev/bus/usb/001/094 test 28,   26.948033 secs
/dev/bus/usb/001/094 test 29,    0.514908 secs
```

From Linux dmesg:
```
[31068.381667] usbtest 1-9:1.0: TEST 0:  NOP
[31068.382166] usbtest 1-9:1.0: TEST 1:  write 1024 bytes 1000 times
[31069.326663] usbtest 1-9:1.0: TEST 2:  read 1024 bytes 1000 times
[31070.218977] usbtest 1-9:1.0: TEST 3:  write/1024 0..1024 bytes 1000 times
[31071.163182] usbtest 1-9:1.0: TEST 4:  read/1024 0..1024 bytes 1000 times
[31072.056458] usbtest 1-9:1.0: TEST 5:  write 1000 sglists 32 entries of 1024 bytes
[31099.061265] usbtest 1-9:1.0: TEST 6:  read 1000 sglists 32 entries of 1024 bytes
[31126.092412] usbtest 1-9:1.0: TEST 7:  write/1024 1000 sglists 32 entries 0..1024 bytes
[31153.163336] usbtest 1-9:1.0: TEST 8:  read/1024 1000 sglists 32 entries 0..1024 bytes
[31180.240162] usbtest 1-9:1.0: TEST 9:  ch9 (subset) control tests, 1000 times
[31180.996691] usbtest 1-9:1.0: TEST 10:  queue 32 control calls, 1000 times
[31184.054873] usbtest 1-9:1.0: TEST 11:  unlink 1000 reads of 1024
[31184.059360] usbtest 1-9:1.0: unlink retry
[31191.403472] usbtest 1-9:1.0: unlink retry
[31191.403481] usbtest 1-9:1.0: unlink retry
[31194.055587] usbtest 1-9:1.0: unlink retry
[31194.723593] usbtest 1-9:1.0: unlink retry
[31196.179822] usbtest 1-9:1.0: unlink retry
[31198.571661] usbtest 1-9:1.0: unlink retry
[31198.571669] usbtest 1-9:1.0: unlink retry
[31198.571673] usbtest 1-9:1.0: unlink retry
[31198.571677] usbtest 1-9:1.0: unlink retry
[31198.571680] usbtest 1-9:1.0: unlink retry
[31198.571683] usbtest 1-9:1.0: unlink retry
[31198.571687] usbtest 1-9:1.0: unlink retry
[31198.571690] usbtest 1-9:1.0: unlink retry
[31201.348461] usbtest 1-9:1.0: TEST 12:  unlink 1000 writes of 1024
[31218.016078] usbtest 1-9:1.0: unlink retry
[31218.016087] usbtest 1-9:1.0: unlink retry
[31218.016091] usbtest 1-9:1.0: unlink retry
[31218.016095] usbtest 1-9:1.0: unlink retry
[31218.948884] usbtest 1-9:1.0: TEST 13:  set/clear 1000 halts
[31229.644546] usb 1-9: verify_not_halted failed, iterations left 0, status -110 (not 0)
[31229.644566] usbtest 1-9:1.0: halts failed, iterations left 862
[31229.645269] usbtest 1-9:1.0: TEST 14:  1000 ep0out, 1..1024 vary 1024
[31229.647266] usbtest 1-9:1.0: TEST 17:  write odd addr 1024 bytes 1000 times core map
[31230.628454] usbtest 1-9:1.0: TEST 18:  read odd addr 1024 bytes 1000 times core map
[31231.615425] usbtest 1-9:1.0: TEST 19:  write odd addr 1024 bytes 1000 times premapped
[31232.600005] usbtest 1-9:1.0: TEST 20:  read odd addr 1024 bytes 1000 times premapped
[31233.578055] usbtest 1-9:1.0: TEST 21:  1000 ep0out odd addr, 1..1024 vary 1024
[31233.579975] usbtest 1-9:1.0: TEST 24:  unlink from 1000 queues of 32 1024-byte writes
[31261.912807] usbtest 1-9:1.0: TEST 27: bulk write 31Mbytes
[31288.861041] usbtest 1-9:1.0: TEST 28: bulk read 31Mbytes
[31315.810286] usbtest 1-9:1.0: TEST 29: Clear toggle between bulk writes 1000 times
```

Details:
MPU Disabled
Zephyr main fadc5d15a3

Running test 13 with MPU enabled a MPU FAULT happen.
NOTE: USB descriptors memory uses SRAM1 space.
```
*** Booting Zephyr OS build zephyr-v2.6.0-196-g7369a2472c13  ***
[00:00:00.000,000] <inf> i2c_sam_twim: Device TWIM_1 initialized
[00:00:00.006,000] <inf> main: entered main.
uart:~$ 
[00:00:19.747,000] <err> os: ***** MPU FAULT *****
[00:00:19.747,000] <err> os:   Data Access Violation
[00:00:19.747,000] <err> os:   MMFAR Address: 0x0
[00:00:19.747,000] <err> os: r0/a1:  0x20000430  r1/a2:  0x200011c8  r2/a3:  0x00000000
[00:00:19.747,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000001 r14/lr:  0x0000beb3
[00:00:19.747,000] <err> os:  xpsr:  0x61000022
[00:00:19.747,000] <err> os: Faulting instruction address (r15/pc): 0x0000f122
[00:00:19.747,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:19.747,000] <err> os: Fault during interrupt handling
[00:00:19.747,000] <err> os: Current thread: 0x20000430 (idle 00)
[00:00:19.822,000] <err> os: Halting system
```

[zephyr.lst.txt](https://github.com/zephyrproject-rtos/zephyr/files/6751721/zephyr.lst.txt)
[zephyr.map.txt](https://github.com/zephyrproject-rtos/zephyr/files/6751722/zephyr.map.txt)

### CODE REVIEW

File: drivers/usb/device/CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-zephyr_sources_ifdef(CONFIG_USB_DW       usb_dc_dw.c)
-zephyr_sources_ifdef(CONFIG_USB_DC_STM32 usb_dc_stm32.c)
-zephyr_sources_ifdef(CONFIG_USB_DC_SAM0  usb_dc_sam0.c)
-zephyr_sources_ifdef(CONFIG_USB_DC_SAM   usb_dc_sam.c)
-zephyr_sources_ifdef(CONFIG_USB_NRFX     usb_dc_nrfx.c)
-zephyr_sources_ifdef(CONFIG_USB_KINETIS  usb_dc_kinetis.c)
+zephyr_sources_ifdef(CONFIG_USB_DW           usb_dc_dw.c)
+zephyr_sources_ifdef(CONFIG_USB_DC_STM32     usb_dc_stm32.c)
+zephyr_sources_ifdef(CONFIG_USB_DC_SAM0      usb_dc_sam0.c)
+zephyr_sources_ifdef(CONFIG_USB_DC_SAM_USBHS usb_dc_sam_usbhs.c)
```
Comment (jfischer-no) [inline drivers/usb/device/CMakeLists.txt:7]: Please move this patch to own PR, this does not draft/belong here. Also the new driver does not seem to be in conflict with usb_dc_sam.c driver.

Comment (nandojve) [reply drivers/usb/device/CMakeLists.txt:7]: Done #36763

**Comment by nandojve** [comment]
Report from Linux testusb:

```
sudo ./testusb -v 512 -D /dev/bus/usb/001/014
./testusb: /dev/bus/usb/001/014 may see only control tests
/dev/bus/usb/001/014 test 0,    0.000017 secs
/dev/bus/usb/001/014 test 1,    0.944672 secs
/dev/bus/usb/001/014 test 2,    0.943044 secs
/dev/bus/usb/001/014 test 3,    0.494024 secs
/dev/bus/usb/001/014 test 4,    0.489585 secs
/dev/bus/usb/001/014 test 5,   27.046488 secs
/dev/bus/usb/001/014 test 6,   27.026396 secs
/dev/bus/usb/001/014 test 7,   13.999698 secs
/dev/bus/usb/001/014 test 8,   13.999526 secs
/dev/bus/usb/001/014 test 9,    0.772943 secs
/dev/bus/usb/001/014 test 10,    3.116524 secs
/dev/bus/usb/001/014 test 11,   16.004006 secs
/dev/bus/usb/001/014 test 12,   16.079510 secs
/dev/bus/usb/001/014 test 13 --> 22 (error 22)
/dev/bus/usb/001/014 test 14,    1.197665 secs
/dev/bus/usb/001/014 test 17,    0.891854 secs
/dev/bus/usb/001/014 test 18,    0.923817 secs
/dev/bus/usb/001/014 test 19,    0.901665 secs
/dev/bus/usb/001/014 test 20,    0.893881 secs
/dev/bus/usb/001/014 test 21,    1.210718 secs
/dev/bus/usb/001/014 test 24,   28.282810 secs
/dev/bus/usb/001/014 test 27,   26.947043 secs
/dev/bus/usb/001/014 test 28,   26.948385 secs
/dev/bus/usb/001/014 test 29,    0.431801 secs
```

From Linux dmesg:

```
[ 2793.426227] usbtest 1-9:1.0: TEST 0:  NOP
[ 2793.426800] usbtest 1-9:1.0: TEST 1:  write 1024 bytes 1000 times
[ 2794.372070] usbtest 1-9:1.0: TEST 2:  read 1024 bytes 1000 times
[ 2795.315735] usbtest 1-9:1.0: TEST 3:  write/512 0..1024 bytes 1000 times
[ 2795.810279] usbtest 1-9:1.0: TEST 4:  read/512 0..1024 bytes 1000 times
[ 2796.300471] usbtest 1-9:1.0: TEST 5:  write 1000 sglists 32 entries of 1024 bytes
[ 2823.348040] usbtest 1-9:1.0: TEST 6:  read 1000 sglists 32 entries of 1024 bytes
[ 2850.375505] usbtest 1-9:1.0: TEST 7:  write/512 1000 sglists 32 entries 0..1024 bytes
[ 2864.376067] usbtest 1-9:1.0: TEST 8:  read/512 1000 sglists 32 entries 0..1024 bytes
[ 2878.376427] usbtest 1-9:1.0: TEST 9:  ch9 (subset) control tests, 1000 times
[ 2879.149828] usbtest 1-9:1.0: TEST 10:  queue 32 control calls, 1000 times
[ 2882.266887] usbtest 1-9:1.0: TEST 11:  unlink 1000 reads of 1024
[ 2898.271701] usbtest 1-9:1.0: TEST 12:  unlink 1000 writes of 1024
[ 2914.352071] usbtest 1-9:1.0: TEST 13:  set/clear 1000 halts
[ 2925.127663] usb 1-9: verify_not_halted failed, iterations left 0, status -110 (not 0)
[ 2925.127667] usbtest 1-9:1.0: halts failed, iterations left 771
[ 2925.128241] usbtest 1-9:1.0: TEST 14:  1000 ep0out, 1..1024 vary 512
[ 2926.327528] usbtest 1-9:1.0: TEST 17:  write odd addr 1024 bytes 1000 times core map
[ 2927.219946] usbtest 1-9:1.0: TEST 18:  read odd addr 1024 bytes 1000 times core map
[ 2928.144390] usbtest 1-9:1.0: TEST 19:  write odd addr 1024 bytes 1000 times premapped
[ 2929.046598] usbtest 1-9:1.0: TEST 20:  read odd addr 1024 bytes 1000 times premapped
[ 2929.941006] usbtest 1-9:1.0: TEST 21:  1000 ep0out odd addr, 1..1024 vary 512
[ 2931.153333] usbtest 1-9:1.0: TEST 24:  unlink from 1000 queues of 32 1024-byte writes
[ 2959.438031] usbtest 1-9:1.0: TEST 27: bulk write 31Mbytes
[ 2986.386058] usbtest 1-9:1.0: TEST 28: bulk read 31Mbytes
[ 3013.335425] usbtest 1-9:1.0: TEST 29: Clear toggle between bulk writes 1000 times
```

Still not complete pass on `TEST 13` but it `halt/un-halt ep 0`. The `samples/subsys/usb/cdc_acm` works well on Windows and Linux.

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 2]
```diff
@@ -0,0 +1,1160 @@
...
+		/* RX - OUT direction */
+		regs->UECONSET[ep_idx] = USBC_UECON0SET_RXOUTES;
+	}
+}
+
+static void usb_dc_ep_isr_sta(uint8_t ep_idx)
+{
+	uint32_t sr = regs->UESTA[ep_idx];
+
+#if defined(CONFIG_USB_DRIVER_LOG_LEVEL_DBG)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c]: Please add usb_dc_ep_isr_sta_dbg() inside `#if defined(CONFIG_USB_DRIVER_LOG_LEVEL_DBG)` at L:25, similar to usb_dc_sam_usbc_clean_sta_dbg()

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 3]
```diff
@@ -0,0 +1,1160 @@
...
+		}
+
+		/* SETUP data received */
+		dev_data.ep_data[0].cb_out(USB_EP_DIR_OUT, USB_DC_EP_SETUP);
+
+		return;
+	}
+
+	if (sr & USBC_UESTA0_RXOUTI) {
+
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c:347]: remove empty line

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 4]
```diff
@@ -0,0 +1,1160 @@
...
+			 * done after the ack to the host completes else the
+			 * ack will get dropped.
+			 */
+			regs->UDCON |= USBC_UDCON_ADDEN;
+		}
+	}
+
+	if (sr & USBC_UESTA0_NAKOUTI) {
+
+		/** Start Control Read State */
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c]: you can remove empty lines above/below, same for `/** Start Control Write State */`

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 5]
```diff
@@ -0,0 +1,1160 @@
...
+}
+
+static void usb_dc_sam_usbc_isr(void)
+{
+	uint32_t sr = regs->UDINT;
+
+#ifdef CONFIG_USB_DEVICE_SOF
+	/* SOF interrupt */
+	if (sr & USBC_UDINT_SOF) {
+		LOG_DBG("ISR: Start of Frame");
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c:463]: Please remove `LOG_DBG("ISR: Start of Frame");`, it will just flood logging

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 6]
```diff
@@ -0,0 +1,1160 @@
...
+	ep_bank = usb_dc_sam_usbc_ep_curr_bank(ep_idx);
+	data_len = dev_desc[ep_bank].SIZES.byte_count;
+
+	/*
+	 * The code below emulates the Quark FIFO which the Zephyr USB
+	 * API is based on.  Reading with buf == NULL returns the
+	 * number of bytes available and starts the read.  The caller
+	 * then keeps calling until all bytes are consumed which
+	 * also marks the OUT buffer as freed.
+	 */
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c:1311]: please remove comment regarding Quark FIFO, current device driver API just expects this behavior.

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 7]
```diff
@@ -0,0 +1,1160 @@
...
+	ep_desc_bk = ((struct sam_usbc_desc_table_t *) &dev_desc)
+		   + (ep_idx * 2);
+	for (int i = ep_idx; i < NUM_OF_EP_MAX; i++) {
+		if (!dev_data.ep_data[i].is_configured && (i != ep_idx)) {
+			ep_desc_bk += 2;
+			continue;
+		}
+
+		/* Alloc bank 0 */
+		ep_desc_bk->ep_pipe_addr = ((uint8_t *) USBC_RAM_ADDR)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc.c:145]: Is that dedicated USB device controller RAM?

File: drivers/usb/device/usb_dc_sam_usbc_priv.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+#define EP_ADDR2IDX(ep)         ((ep) & ~USB_EP_DIR_MASK)
+#define EP_ADDR2DIR(ep)         ((ep) & USB_EP_DIR_MASK)
+#define EP_UDINT_MASK()         (0x000FF000)
+
+#define NUM_OF_EP_MAX           DT_INST_PROP(0, num_bidir_endpoints)
+#define USBC_RAM_ADDR           DT_REG_ADDR(DT_NODELABEL(sram1))
+#define USBC_RAM_SIZE           DT_REG_SIZE(DT_NODELABEL(sram1))
+#define USBC_is_frozen_clk()    (USBC->USBCON & USBC_USBCON_FRZCLK)
+#define USBC_freeze_clk()       (USBC->USBCON |= USBC_USBCON_FRZCLK)
+#define USBC_unfreeze_clk()     do {					     \
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc_priv.h:17]: Why can it not be a (maybe with ALWAYS_INLINE attribute) function? (same for all this frozen_clk macros)

File: drivers/usb/device/usb_dc_sam_usbc_priv.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc_priv.h:1]: You can just move everything that is here to usb_dc_sam_usbc.c, it is not that much.

File: drivers/usb/device/usb_dc_sam_usbc_priv.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ * Copyright (c) 2020 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Helper macros to make it easier to work with endpoint numbers */
+#define EP_ADDR2IDX(ep)         ((ep) & ~USB_EP_DIR_MASK)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc_priv.h:8]: please use USB_EP_GET_IDX(ep) macro from include/drivers/usb/usb_dc.h

File: drivers/usb/device/usb_dc_sam_usbc_priv.h [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+#define EP_ADDR2DIR(ep)         ((ep) & USB_EP_DIR_MASK)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_sam_usbc_priv.h:9]: please use USB_EP_DIR_IS_IN(ep) macro from include/drivers/usb/usb_dc.h

**Review by jfischer-no — COMMENTED** [review]
Please rebase on latest main.

File: dts/bindings/usb/atmel,sam-usbc.yaml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    interrupts:
+      required: true
+
+    peripheral-id:
+      type: int
+      required: true
+      description: peripheral ID
+
+    pinctrl-0:
```
[the remark this answers is not in the document]
Comment (nandojve) [reply dts/bindings/usb/atmel,sam-usbc.yaml:24]: Yes, it is. This pins are shared with USART2 and CATB module.

File: drivers/usb/device/usb_dc_sam_usbc.c (hunk 7 above)
Comment (nandojve) [reply drivers/usb/device/usb_dc_sam_usbc.c:145]: No, USB controller have their own dedicated DMA and can access that memory. That was dedicated to USB stack on Zephyr.

https://github.com/zephyrproject-rtos/zephyr/blob/f5a3deaad382b0cb522cbfd516a9b44eb7ca3311/dts/arm/atmel/sam4l.dtsi#L42-L51

File: drivers/usb/device/usb_dc_sam_usbc.c [hunk 13]
```diff
@@ -0,0 +1,1160 @@
...
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_dc_sam_usbc, CONFIG_USB_DRIVER_LOG_LEVEL);
+
+#include <kernel.h>
+#include <usb/usb_device.h>
+#include <soc.h>
+#include <string.h>
+
+#include "usb_dc_sam_usbc_priv.h"
```
[the remark this answers is not in the document]
Comment (nandojve) [reply drivers/usb/device/usb_dc_sam_usbc.c:18]: Moved to c file.

**Comment by nandojve** [comment]
Hi @jfischer-no ,

Thank you for your review and USB stack improvements. Current version is full passing on Linux USB tests.

```
sudo ./testusb -v 512 -D /dev/bus/usb/001/032
./testusb: /dev/bus/usb/001/032 may see only control tests
/dev/bus/usb/001/032 test 0,    0.000002 secs
/dev/bus/usb/001/032 test 1,    0.916624 secs
/dev/bus/usb/001/032 test 2,    0.929512 secs
/dev/bus/usb/001/032 test 3,    0.486187 secs
/dev/bus/usb/001/032 test 4,    0.489856 secs
/dev/bus/usb/001/032 test 5,   28.474857 secs
/dev/bus/usb/001/032 test 6,   27.046195 secs
/dev/bus/usb/001/032 test 7,   14.734015 secs
/dev/bus/usb/001/032 test 8,   14.004437 secs
/dev/bus/usb/001/032 test 9,    0.925995 secs
/dev/bus/usb/001/032 test 10,    3.170833 secs
/dev/bus/usb/001/032 test 11,   17.682407 secs
/dev/bus/usb/001/032 test 12,   18.007703 secs
/dev/bus/usb/001/032 test 13,    4.128832 secs
/dev/bus/usb/001/032 test 14,    1.340714 secs
/dev/bus/usb/001/032 test 17,    1.019905 secs
/dev/bus/usb/001/032 test 18,    1.013034 secs
/dev/bus/usb/001/032 test 19,    1.017387 secs
/dev/bus/usb/001/032 test 20,    0.996685 secs
/dev/bus/usb/001/032 test 21,    1.326164 secs
/dev/bus/usb/001/032 test 24,   28.166526 secs
/dev/bus/usb/001/032 test 27,   28.320512 secs
/dev/bus/usb/001/032 test 28,   26.950037 secs
/dev/bus/usb/001/032 test 29,    0.439828 secs
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Not tested, also had no time to look in datasheet, looks okay, just a few comments. You can squash second and third commit if you like.

**Review by jfischer-no — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
